### PR TITLE
references by uuid

### PR DESCRIPTION
**Pull request by ElectricMaxxx**
This should wrap up #493 

didn't implement the stuff, commented in code not to change.

But there are still the [Questions](https://github.com/doctrine/phpcr-odm/pull/493#issuecomment-43854355) of @dbu

I go top to button through the text/conversation, cause it is really hard to remember what the purpose was, even for me :-)

### CODE REVIEW

File: lib/Doctrine/ODM/PHPCR/DocumentManager.php [hunk 1]
```diff
@@ -917,13 +917,11 @@ public function initializeObject($document)
      */
     public function getNodeForDocument($document)
     {
-        if (!is_object($document)) {
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/DocumentManager.php]: Why do we remove this validation?

File: lib/Doctrine/ODM/PHPCR/DocumentManager.php [hunk 2]
```diff
@@ -917,13 +917,11 @@ public function initializeObject($document)
...
-            throw new InvalidArgumentException('Parameter $document needs to be an object, '.gettype($document).' given');
+        if (!$identifier = $this->unitOfWork->getDocumentId($document)) {
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/DocumentManager.php]: `if (null === $identifier = $this->...`

File: lib/Doctrine/ODM/PHPCR/DocumentManager.php [hunk 3]
```diff
@@ -917,13 +917,11 @@ public function initializeObject($document)
...
+            throw new InvalidArgumentException('This document is not managed by this manager.');
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/DocumentManager.php]: `sprintf('Document of class "%s" is not managed', get_class($document))`

File: lib/Doctrine/ODM/PHPCR/UnitOfWork.php [hunk 4]
```diff
@@ -674,7 +675,7 @@ public function getOrCreateProxy($targetId, $className, $locale = null)
         $metadata      = $this->dm->getClassMetadata($className);
         $proxyDocument = $this->dm->getProxyFactory()->getProxy($className, array($metadata->identifier => $targetId));
 
-        // register the document under its own id
+        // register the document under its own id/uuid
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/UnitOfWork.php]: `path/uuid`.

File: lib/Doctrine/ODM/PHPCR/UnitOfWork.php [hunk 5]
```diff
@@ -3113,6 +3126,20 @@ public function getDocumentById($id)
     }
 
     /**
+     * Creates a node from a given path or an uuid
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/UnitOfWork.php]: `given path or UUID`

File: lib/Doctrine/ODM/PHPCR/UnitOfWork.php [hunk 6]
```diff
@@ -3063,18 +3069,25 @@ private function unregisterDocument($document)
...
 
         // frozen nodes need another state so they are managed but not included for updates
-        $frozen = $this->session->nodeExists($id) && $this->session->getNode($id)->isNodeType('nt:frozenNode');
+        try {
+            $node = $this->getNodeByPathOrUuid($pathOrUuid);
+            $frozen = $node instanceof  NodeInterface ? $node->isNodeType('nt:frozenNode') : false;
+        } catch (ItemNotFoundException $e) {
+            $frozen = false;
+        } catch (PathNotFoundException $e) {
+            $frozen = false;
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/UnitOfWork.php]: `$frozen = false` can be moved before the `try` block (doesn't need to be repeated twice)

File: tests/Doctrine/Tests/ODM/PHPCR/UnitOfWorkTest.php [hunk 7]
```diff
@@ -74,54 +75,66 @@ public function setUp()
...
             ->with($id)
             ->will($this->returnValue(true));
-        
+
         $nodeData = array(
             "jcr:primaryType" => $primaryType,
             "jcr:system" => array(),
             'username' => $username,
         );
+        if (null !== $uuid) {
```
Comment (dantleech) [inline tests/Doctrine/Tests/ODM/PHPCR/UnitOfWorkTest.php]: I prefer a blank line before most if statements.

File: tests/Doctrine/Tests/ODM/PHPCR/UnitOfWorkTest.php [hunk 8]
```diff
@@ -74,54 +75,66 @@ public function setUp()
...
-        
+
+        if (null !== $uuid) {
+            $this->session->expects($this->any())
+                ->method('getNodeByIdentifier')
+                ->with($uuid)
+                ->will($this->returnValue($node));
+        }
+
+
```
Comment (dantleech) [inline tests/Doctrine/Tests/ODM/PHPCR/UnitOfWorkTest.php]: One blank line too many

**Review by dantleech — COMMENTED** [review]
Can you summarize the motivation behind this PR in the description? I don't really know the full implications of this change.

**Comment by ElectricMaxxx** [comment]
This will fail for sure now ... Did a lot of changes

* introduced a object hash - uuid map, which is filled when a document is registered by its uuid (on `DocumentManager::getReference()` call with proxies referencing documents by their uuid <- @dantleech purpose)
* introduced a uuid - path map to find documents by their path in identity map, which are managed by uuid

so all documents still live in `identityMap` by their id, but the id can be the absolute paths directly or uuids which needs to be converted into paths (uuid - id map).

File: lib/Doctrine/ODM/PHPCR/UnitOfWork.php [hunk 9]
```diff
@@ -100,6 +101,18 @@ class UnitOfWork
     private $documentIds = array();
 
     /**
+     * @var array
+     */
+    private $documentUuids = array();
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/UnitOfWork.php]: Ideally this would be better named `$oidToUuidMap`, but to be consistent this naming is fine for now I guess. But it should be commented.

File: lib/Doctrine/ODM/PHPCR/UnitOfWork.php [hunk 10]
```diff
@@ -100,6 +101,18 @@ class UnitOfWork
...
+
+    /**
+     * Mapping to find ids (abs. path) of document managed by its uuid.
+     *
+     * @var array
+     */
+    private $uuidIdMapping = array();
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/UnitOfWork.php]: `Id` in `uuidId` is redundant. Better to name `$uuidToPathMap`.

File: lib/Doctrine/ODM/PHPCR/UnitOfWork.php [hunk 11]
```diff
@@ -3058,23 +3076,43 @@ private function unregisterDocument($document)
             $this->documentVersion[$oid]
         );
 
+        if (!empty($this->documentUuids[$oid]) && array_key_exists($this->documentUuids[$oid], $this->uuidIdMapping)) {
```
Comment (dantleech) [inline lib/Doctrine/ODM/PHPCR/UnitOfWork.php]: Why check for `!empty`?

**Review by dantleech — COMMENTED** [review]
Still not sure what the *motivation* of this is? Regarding the mapping I wonder if it is worth introducing a Registry class to encapsulate the various maps, reducing complication in this already very complicated class, something *like*: https://github.com/sulu/sulu-document-manager/blob/master/lib/DocumentRegistry.php.

But maybe something for a later iteration.
